src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs

```diff
@@ -108,7 +121,48 @@ public void Apply(Workspace workspace, Document fromDocument, IEnumerable<CodeAc
             var oldSolution = workspace.CurrentSolution;
             Solution updatedSolution = oldSolution;
 
-            foreach (var operation in operations)
+            var operationsList = operations.ToList();
+            if (operationsList.Count > 1)
+            {
+                // Make a linked undo to wrap all these operations.  This way we should
+                // be able to undo them all with one user action.
+                using (var transaction = workspace.OpenGlobalUndoTransaction(title))
```

Comment:
This is only for code actions with *multiple* operations.  So the 'simple' fix that only changes a single document won't end up going through this code path.

note that multi-operation code actions are already rare.  It effectively means "we couldn't express the change as a solution to solution change, and we needed to do additional work beyond that".  In that case, i think it's appropriate to actually wrap that in an undo.

I'm happy to comment the code as such if you like.